### ISSUE TITLE
Virtual String Tree CopyToClipboard does not copy cells colors

### ISSUE THREAD

**Issue by Michel-WB-DivByZero** [JAM-Software/Virtual-TreeView#588]
Hi,
We just upgraded Embarcadero Seattle and to the 6.2 version of Virtual String Tree and the CopyToClipboard does not copy ht e grid cells color to the clipboard.

Also, with a grid that contains about 150 rows with colors, accessing the RTF format in the clipboard results in an access violation in the application.

Could you help me with this issue?

Thanks a lot!

Here is the code that shows this behaviour : 
# Unit1.h

``` javascript
//---------------------------------------------------------------------------

#ifndef Unit1H
#define Unit1H
//---------------------------------------------------------------------------
#include <System.Classes.hpp>
#include <Vcl.Controls.hpp>
#include <Vcl.StdCtrls.hpp>
#include <Vcl.Forms.hpp>
#include "AdvGrid.hpp"
#include "AdvObj.hpp"
#include "BaseGrid.hpp"
#include <Vcl.Grids.hpp>
#include "VirtualTrees.hpp"
//---------------------------------------------------------------------------
struct vstItem
{
  String col0;
  String col1;
  String col2;
  String col3;
  String col4;
  String col5;

  int irow;
};

class TForm1 : public TForm
{
__published:    // Composants gérés par l'EDI
  TVirtualStringTree *vst;
  TButton *Button1;
  void __fastcall FormShow(TObject *Sender);
  void __fastcall vstGetText(TBaseVirtualTree *Sender, PVirtualNode Node, TColumnIndex Column,
          TVSTTextType TextType, UnicodeString &CellText);
  void __fastcall vstBeforeCellPaint(TBaseVirtualTree *Sender, TCanvas *TargetCanvas,
          PVirtualNode Node, TColumnIndex Column, TVTCellPaintMode CellPaintMode,
          TRect &CellRect, TRect &ContentRect);
  void __fastcall vstPaintText(TBaseVirtualTree *Sender, TCanvas * const TargetCanvas,
          PVirtualNode Node, TColumnIndex Column, TVSTTextType TextType);
  void __fastcall Button1Click(TObject *Sender);

private:    // Déclarations utilisateur
public:     // Déclarations utilisateur
  __fastcall TForm1(TComponent* Owner);
};
//---------------------------------------------------------------------------
extern PACKAGE TForm1 *Form1;
//---------------------------------------------------------------------------
#endif
```
# Unit1.cpp

``` javascript
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include <string>
#include "VirtualTrees.ClipBoard.hpp"

#include "Unit1.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "AdvGrid"
#pragma link "AdvObj"
#pragma link "BaseGrid"
#pragma link "VirtualTrees"
#pragma resource "*.dfm"
TForm1 *Form1;
//---------------------------------------------------------------------------
__fastcall TForm1::TForm1(TComponent* Owner)
  : TForm(Owner)
{
  vst->NodeDataSize = sizeof( vstItem );
}
//---------------------------------------------------------------------------
void __fastcall TForm1::FormShow(TObject *Sender)
{
  vst->Header->Columns->BeginUpdate();

  vst->Clear();

  for ( int i = 0 ; i < 6 ; ++i )
  {
    TVirtualTreeColumn *col = vst->Header->Columns->Add();
    col->Text  = "Col_" + String( i );
    col->Width = 100;

    if ( i == 3 )                                  // test hiding 1 column
      col->Options >> TVTColumnOption::coVisible;
  }

  vst->Header->Background   = clGray;
  vst->Header->Font->Size   = 10;
  vst->Header->Font->Style << fsBold;

  vst->Header->Columns->EndUpdate();

  vst->BeginUpdate();

  for ( int i = 0 ; i < 7 ; ++i )
  {
    TVirtualNode *node = vst->AddChild( NULL );
    vstItem      *item = static_cast<vstItem*>( vst->GetNodeData( node ) );

    char cbase = 'A';
    char *crow = &cbase;

    item->irow = i;

    item->col0 = "Cell_" + String( (*crow)++ ) + "_" + String( i );
    item->col1 = "Cell_" + String( (*crow)++ ) + "_" + String( i );
    item->col2 = "Cell_" + String( (*crow)++ ) + "_" + String( i );
    item->col3 = "Cell_" + String( (*crow)++ ) + "_" + String( i );
    item->col4 = "Cell_" + String( (*crow)++ ) + "_" + String( i );
    item->col5 = "Cell_" + String( (*crow)++ ) + "_" + String( i );
  }

  vst->EndUpdate();
  vst->FullExpand();
}
//---------------------------------------------------------------------------
void __fastcall TForm1::vstGetText(TBaseVirtualTree *Sender, PVirtualNode Node,
                                   TColumnIndex Column, TVSTTextType TextType, UnicodeString &CellText)

{
  vstItem *item = static_cast<vstItem*>( vst->GetNodeData( Node ) );

  if ( Column == 0 )
    CellText = item->col0;
  else if ( Column == 1 )
    CellText = item->col1;
  else if ( Column == 2 )
    CellText = item->col2;
  else if ( Column == 3 )
    CellText = item->col3;
  else if ( Column == 4 )
    CellText = item->col4;
  else if ( Column == 5 )
    CellText = item->col5;
}
//---------------------------------------------------------------------------
void __fastcall TForm1::vstBeforeCellPaint(TBaseVirtualTree *Sender, TCanvas *TargetCanvas,
          PVirtualNode Node, TColumnIndex Column, TVTCellPaintMode CellPaintMode,
          TRect &CellRect, TRect &ContentRect)
{
  vstItem *item = static_cast<vstItem*>( vst->GetNodeData( Node ) );

  if ( item->irow % 2 && Column == 2 )
    TargetCanvas->Brush->Color = clSkyBlue;

  TargetCanvas->FillRect( CellRect );
}
//---------------------------------------------------------------------------
void __fastcall TForm1::vstPaintText(TBaseVirtualTree *Sender, TCanvas * const TargetCanvas,
          PVirtualNode Node, TColumnIndex Column, TVSTTextType TextType)

{
  vstItem *item = static_cast<vstItem*>( vst->GetNodeData( Node ) );

  if ( item->irow % 3 && Column == 3 )
    TargetCanvas->Font->Color = clRed;
  else if ( item->irow % 2 && Column == 4 )
    TargetCanvas->Font->Color = clBlue;
  else if ( item->irow % 2 )
    TargetCanvas->Font->Color = clMedGray;
}
//---------------------------------------------------------------------------

void __fastcall TForm1::Button1Click(TObject *Sender)
{
  try
  {
    Cursor = crHourGlass;
    vst->SelectAll( false );
    vst->CopyToClipboard();
    vst->ClearSelection();
  }
  __finally
  {
    Cursor = crDefault;
  }
}
//---------------------------------------------------------------------------
```
# Unit1.dfm

``` javascript
object Form1: TForm1
  Left = 0
  Top = 0
  Caption = 'Form1'
  ClientHeight = 325
  ClientWidth = 1072
  Color = clBtnFace
  Font.Charset = DEFAULT_CHARSET
  Font.Color = clWindowText
  Font.Height = -11
  Font.Name = 'Tahoma'
  Font.Style = []
  OldCreateOrder = False
  OnShow = FormShow
  PixelsPerInch = 96
  TextHeight = 13
  object vst: TVirtualStringTree
    Left = 0
    Top = 39
    Width = 1072
    Height = 286
    Align = alBottom
    Anchors = [akLeft, akTop, akRight, akBottom]
    CheckImageKind = ckXP
    ClipboardFormats.Strings = (
      'CSV'
      'HTML Format'
      'Plain text'
      'Rich Text Format'
      'Rich Text Format Without Objects'
      'Unicode text'
      'Virtual Tree Data')
    Ctl3D = False
    DragOperations = []
    Header.AutoSizeIndex = 0
    Header.Background = clGray
    Header.Font.Charset = DEFAULT_CHARSET
    Header.Font.Color = clMaroon
    Header.Font.Height = -13
    Header.Font.Name = 'Tahoma'
    Header.Font.Style = [fsBold]
    Header.MainColumn = -1
    Header.Options = [hoColumnResize, hoDblClickResize, hoShowHint, hoShowImages, hoVisible, hoFullRepaintOnResize]
    LineStyle = lsSolid
    ParentCtl3D = False
    ParentShowHint = False
    ScrollBarOptions.AlwaysVisible = True
    SelectionCurveRadius = 5
    ShowHint = False
    TabOrder = 0
    TreeOptions.AutoOptions = [toAutoScrollOnExpand, toAutoSpanColumns, toAutoTristateTracking, toAutoDeleteMovedNodes]
    TreeOptions.MiscOptions = [toAcceptOLEDrop, toFullRepaintOnResize, toInitOnSave, toToggleOnDblClick, toWheelPanning, toVariableNodeHeight]
    TreeOptions.PaintOptions = [toShowDropmark, toThemeAware, toUseBlendedImages]
    TreeOptions.SelectionOptions = [toExtendedFocus, toMultiSelect]
    TreeOptions.StringOptions = [toSaveCaptions]
    OnBeforeCellPaint = vstBeforeCellPaint
    OnGetText = vstGetText
    OnPaintText = vstPaintText
    Columns = <>
    WideDefaultText = ''
  end
  object Button1: TButton
    Left = 32
    Top = 8
    Width = 129
    Height = 25
    Caption = 'copy to clipboard'
    TabOrder = 1
    OnClick = Button1Click
  end
end
```

**Comment by joachimmarder** [comment]
Hi.

It would be best to open an issue at:
https://github.com/Virtual-TreeView/Virtual-TreeView/issues

To which application did you paste the copied content?
Which version of Virtual Treeview were you using before?

Best regards,

Joachim

2015-10-28 11:46 GMT+01:00 Michel-WB-DivByZero notifications@github.com:

**Comment by Michel-WB-DivByZero** [comment]
Hi Joachim,

Thanks for your answer.

We previously used Virtual Treeview versions 4.7, 5.41 and 6.1 on RAD XE, RAD XE6 and RAD XE8.
I tried pasting to MS Excel, MS Word, MS Outlook, WordPad, Notepad++ and LibreOffice spredsheet and word processor.

We tend to think that the change from CopyToClipBoard to CopyToClipboard also implied changes in the implementation of that feature.
BTW: isn't https://github.com/Virtual-TreeView/Virtual-TreeView/issues this page?

Best Regards,

Michel

**Comment by joachimmarder** [comment]
Hi Michel.

> BTW: isn't https://github.com/Virtual-TreeView/Virtual-TreeView/issues
> this page?

Sorry, I missed that your mail was genrated by GitHub and was not a
personal mail.

> We previously used Virtual Treeview versions 4.7, 5.41 and 6.1 on RAD XE,
> RAD XE6 and RAD XE8.

And they were all working as expected?

Best regards,

Joachim

2015-10-30 8:53 GMT+01:00 Michel-WB-DivByZero notifications@github.com:

**Comment by joachimmarder** [comment]
It seems that the HTML clipboard format is broken after the full transition to unicode-strings. The HTML clipboard format does not work with Unicode UTF16, only with UTF8 encoded HTML

**Comment by joachimmarder** [comment]
The HTML clipboard format is now working again. However I don't see any cell background colors with the current source as well as with V5.4.1 using the grid demo from the Advanced demo project.

**Comment by joachimmarder** [comment]
> Also, with a grid that contains about 150 rows with colors, 
> accessing the RTF format in the clipboard results in an access violation in the application.

I was unable to reproduce this with the latest source and your sample app after increasing number of rows to 700.

**Comment by Michel-WB-DivByZero** [comment]
Hi Joachim,
I could not reproduce  the +150 rows issue on a clean environment.
I checked that on the previous version, we actually do not have the cells colors, but still the header row backfround is geyed out. Since the main CopyToClipboard functionality is fixed now, this would be a 'nice to have', not a showstopper.
I understand that your fix will be released with the 6.2.1 version, is that correct?
Thanks a lot for your help!
Michel

**Comment by joachimmarder** [comment]
Supporting cell background color is not an easy task, because it is achieved through custom drawing in event handlers, and unlike Font.Color not persisted anywhere.

The fix for the HTML clipboard format will be included in V6.2.1.
